File: Binary Search Tree/balancedBinaryTree.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
#include "binarySearchTree.hpp"

using namespace std;

// Solution: https://www.youtube.com/watch?v=LU4fGD-fgJQ

class Solution
{
private:
    int findHeight(TreeNode *it)
    {
        if (it == nullptr)
            return -1;

        int l = findHeight(it->left);
        int r = findHeight(it->right);

        return max(l, r) + 1;
    }

public:
    bool isBalanced(TreeNode *root)
    {
        if (root == nullptr)
            return true;

        if (abs(findHeight(root->left) - findHeight(root->right)) > 1)
            return false;

        bool l = isBalanced(root->left);
        bool r = isBalanced(root->right);

        return l && r;
    }
};

int main()
{

    return 0;
}

File: NEETCODE/Trees/balancedBinaryTree.py
from turtle import right
from typing import Optional
from binarySearchTree import TreeNode
from binarySearchTree import BinarySearchTree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def height_dfs(self, root: Optional[TreeNode]) -> int:
        if not root:
            return -1
        return 1 + max(self.height_dfs(root.left), self.height_dfs(root.right))

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        l_height = self.height_dfs(root.left)
        r_height = self.height_dfs(root.right)

        if abs(l_height - r_height) > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)


if __name__ == "__main__":
    tree = BinarySearchTree()
    tree.insert(1)
    tree.insert(2)
    tree.insert(3)

    obj = Solution()
    print(obj.isBalanced(tree.root))
